### ISSUE TITLE
Missing opt.zero_grad() when bfloat16 is enabled causes silent gradient accumulation

### ISSUE THREAD

**Issue by kdwonn** [facebookresearch/nwm#16]
Hi, thanks for the code release! I've really enjoyed reading the paper.

I've discovered a bug in the training loop when using bfloat16 mixed precision training. The `opt.zero_grad()` call is missing in the mixed precision code path, causing unintended gradient accumulation.

https://github.com/facebookresearch/nwm/blob/1bc6af63438678f445f102380f98823c8274797d/train.py#L300

In the `train_step` function, the optimizer gradients are only zeroed when `bfloat_enable=False:`.

Can you give a check? @amirbar
